phper/src/builder/data_type.rs

```diff
@@ -0,0 +1,71 @@
+pub enum ArrayKey {
+    String,
+    Int,
+    Mixed,
+}
+
+pub enum AtomicType {
+    Bool,
+    Int,
+    Float,
+    String,
+    NonEmptyString,
+    IntRange(i64, i64),
+    ClassString,
+    ClassStringOf(String),
+    TraitString,
+    EnumString,
+    CallableString,
+    NumericString,
+    LiteralString,
+    LiteralInt,
+    Numeric,
+    Scalar,
+    Fiber,
+    Object,
+    ObjectWithProperties(Vec<(String, AtomicType)>),
+    ObjectOf(String),
+    List(Box<AtomicType>),
+    Array(Box<(ArrayKey, AtomicType)>),
+    NonEmptyList(Box<AtomicType>),
+    NonEmptyArray(Box<(ArrayKey, AtomicType)>),
+    Shape(Vec<(ArrayKey, AtomicType)>),
+    Iterable(Box<(ArrayKey, AtomicType)>),
+    CallableArray,
+    Callable,
+    Null,
+    True,
+    False,
+    Resource,
+    ClosedResource,
+    Mixed,
+    Union(Vec<AtomicType>),
+}
```

Comment:
see: https://psalm.dev/docs/annotating_code/type_syntax/atomic_types/

when generating code: simplify the type to it's PHP form
when generating stubs: simply the type to it's PHP representation in the signature, but add type atomic type representation in the doc block.

Reply:
note: removed `Fiber`, while we should support Fibers, this PR is not about that, but we need to figure out a way support fibers.